src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Integer.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal ref partial struct CborValueReader
+    {
+        // Implements major type 0 decoding per https://tools.ietf.org/html/rfc7049#section-2.1
+        public ulong ReadUInt64()
+        {
+            CborDataItem header = ReadInitialByte();
+
+            switch (header.MajorType)
+            {
+                case CborMajorType.UnsignedInteger:
+                    ulong value = ReadUnsignedInteger(header, out int additionalBytes);
+                    AdvanceBuffer(1 + additionalBytes);
+                    return value;
+
+                default:
+                    throw new InvalidOperationException("Data item type mismatch");
+            }
+        }
+
+        public bool TryReadUInt64(out ulong value)
+        {
+            CborDataItem header = ReadInitialByte();
+
+            switch (header.MajorType)
+            {
+                case CborMajorType.UnsignedInteger:
+                    value = ReadUnsignedInteger(header, out int additionalBytes);
+                    AdvanceBuffer(1 + additionalBytes);
+                    return true;
+
+                default:
+                    value = 0;
+                    return false;
```

Comment:
We should also test the TryReads, for both true and false.

Reply:
I skipped testing here, as I'm not 100% sure what the correct TryRead* semantics should be at this point. I didn't include TryRead methods for the string types and contemplated removing integer TryReads for the context of this PR. Perhaps I should go ahead and remove?

Reply:
> I'm not 100% sure what the correct TryRead* semantics should be at this point

The only semantics that make sense to me are for the integers as out of range (matches what we did in Utf8JsonReader.TryGetInt32).

TryReadInt32:

* Current value type isn't +int or -int? throw InvalidOperationException.
* Current value is outside [int.MinValue, int.MaxValue]? value = 0, return false.
* Else value = parsedValue, return true.

> I didn't include TryRead methods for the string types

I don't think that a TryRead makes sense for them, again, basing things on decisions we made for the JSON reader in 3.

> Perhaps I should go ahead and remove?

They'll probably get defined at some point, and when they get defined they need tests. There's a possibility that the semantics will change at some point, but the test can be changed, too.

Reply:
> There's a possibility that the semantics will change at some point, but the test can be changed, too.

Sure, but there's always a cost associated with changing tests guarding against an ill-conceived design. I don't mind doing it like this though.

Reply:
> I don't think that a TryRead makes sense for them, again, basing things on decisions we made for the JSON reader in 3.

Perhaps in the context of writing the result to a buffer supplied by the caller?

Reply:
Fair enough :) That TryRead is, amusingly, a Try-Write, and so it would then get Try-Write semantics (in both that string sense and the int sense the false means "it's the right type, but it doesn't fit")

Reply:
> Current value type isn't +int or -int? throw InvalidOperationException.

I'm not too sure about that one. I get that `false` should represent a single failure mode, but both that and value falling out of range could be vaguely equated as casting errors. In practice it would prompt users to almost always resort to `Peek()` calls beforehand.

Reply:
>  In practice it would prompt users to almost always resort to Peek() calls beforehand.

For a program doing dynamic work / particle emission, sure.  But when the data comes from a schema, I don't think that'd be common. Like, when the schema says something is an array of integer values you're only looking for "does this fit in Int32, or do I need something bigger?".

Maybe the fact that CBOR map keys can be any value means that all map processing is effectively dynamic, which could justify being different from what we did for Utf8JsonReader.  Really it comes down to writing realistic sample code and seeing what felt right.  With JSON and DER the samples kept coming up as "I know the type this needs to be" (or "I have to ask what type it is, because I need to internally do something different if it's a string or a number... now I know what type it is before I do further work"); there weren't realistic-looking samples (other than particle emitters) that would have just Try-Gotten everything and seen what stuck.

Reply:
> For a program doing dynamic work / particle emission, sure. But when the data comes from a schema, I don't think that'd be common.

I suppose this begs the question whether TryRead* methods for integers would be a useful addition altogether. You wouldn't need it for a schema (provided it doesn't have unions), and everything else could be handled canonically using the `Peek()` method. 

What's more, if the only utility of a TryRead*Int* method is to guard against integer overflows, then I would probably say it's of low value vis-a-vis the api complexity it introduces. Assuming your application has issues with integer overflows, then they could be easily worked around using `Peek()`.

Reply:
If you don't think they'll be useful, we can leave them out.